cpp/src/arrow/compute/kernels/codegen_internal.cc

```diff
@@ -194,6 +194,11 @@ Result<ValueDescr> FirstType(KernelContext*, const std::vector<ValueDescr>& desc
   return result;
 }
 
+Result<ValueDescr> ListValuesType(KernelContext*, const std::vector<ValueDescr>& args) {
+  const auto& list_type = checked_cast<const BaseListType&>(*args[0].type);
+  return ValueDescr::Array(list_type.value_type());
```

Comment:
It might be good to use GetBroadcastShape as FirstType does above so that this can handle all-scalar inputs too.

Reply:
Thanks @lidavidm
I just moved this code from other place, although I'm not familiar with this part I'll be happy to apply any improvement. So, could you please elaborate how I can use GetBroadcastShape here?

Reply:
You can see it above: https://github.com/apache/arrow/blob/954ca3e0962a45f588c1ab19fd100c728b3d7c42/cpp/src/arrow/compute/kernels/codegen_internal.cc#L191-L195

It would be something like this:

```cpp
return ValueDescr(list_type.value_type(), GetBroadcastShape(args));
```

Note you may want to then update this kernel to support scalar inputs. (If you want to use CheckScalar for testing, you'll need to.)

Reply:
Thanks @lidavidm !
What would be the behavior for this function in the case of scalar inputs?
Can we move this part of the discussion into the [jira ticket](https://issues.apache.org/jira/browse/ARROW-12669?focusedCommentId=17415119&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17415119)? Maybe you can draw some examples there! cc @ianmcook

Reply:
I don't think there's anything complicated. `list_element([1, 2, 3], 1) == Scalar(2)`. The idea behind a scalar function is that the scalar behavior should follow from the array behavior; it's effectively as if you had a length-1 array, except you accept a scalar and return a scalar.

Reply:
Thanks David, indeed, I don't think it is complicated. I just want to validate if the behavior you have in mind is aligned with the requested implementation in the ticket.

Reply:
@lidavidm [here](https://issues.apache.org/jira/browse/ARROW-12669?focusedCommentId=17416195&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17416195) thanks!